plugins/ltac/tacinterp.ml

```diff
@@ -162,17 +162,26 @@ let catching_error call_trace fail (e, info) =
     fail located_exc
   end
 
-let catch_error call_trace f x =
+let locate_if_not_already ?loc (e, info) =
+  match Loc.get_loc info with
+  | None   -> (e, Option.cata (Loc.add_loc info) info loc)
+  | Some l -> (e, info)
+
+let catch_error ?loc call_trace f x =
   try f x
   with e when CErrors.noncritical e ->
     let e = Exninfo.capture e in
+    let e = locate_if_not_already ?loc e in
     catching_error call_trace Exninfo.iraise e
 
-let wrap_error tac k =
-  if is_traced () then Proofview.tclORELSE tac k else tac
+let wrap_error ?loc tac k =
+  if is_traced () then Proofview.tclORELSE tac k
```

Comment:
Actually I wonder why we use `tclORELSE` at all in this case. Apart from having an involved, slow implementation, it is also likely to change the semantics w.r.t. backtracking when tracing. While you're at it I think it'd be even better to change it into a `tclOR`, and even restrict the type of `k` to `exninfo -> exninfo`.

Reply:
I moved the `tclORELSE` to a `tclERROR`. Changing the type of `k` would require more thinking and I'm also not fully sure of how you would like it to be done exactly. Don't hesitate to push a proposal to the branch.